Add ffi::OsString and OsStr

aturon opened the pull request:
Per [RFC 517](https://github.com/rust-lang/rfcs/pull/575/), this commit introduces platform-native strings. The API is essentially as described in the RFC.

The WTF-8 implementation is adapted from @SimonSapin's [implementation](https://github.com/SimonSapin/rust-wtf8). To make this work, some encodign and decoding functionality in `libcore` is now exported in a "raw" fashion reusable for WTF-8. These exports are *not* reexported in `std`, nor are they stable.

rust-highfive commented:
r? @alexcrichton

(rust_highfive has picked a reviewer for you, use r? to override)

aturon commented:
Note, this is work in progress and should not be merged until the [RFC](https://github.com/rust-lang/rfcs/pull/575/) is accepted.

cc @SimonSapin

alexcrichton commented on src/libstd/ffi/os_str.rs:
```diff
@@ -0,0 +1,214 @@
+    /// Yield a `&str` slice if the `OsStr` is valid unicode.
+    ///
+    /// This conversion may entail doing a check for UTF-8 validity.
+    pub fn to_str(&self) -> Option<&str> {
```
to_str => as_str?

alexcrichton replied:
Oh right, this can do work (UTF-8 validation).

tbu- replied:
What use is this function? Either you want a string to display, or you don't care about a string. What does it help if you get an `Option<&str>` except for errornous code that assumes that all pathes are representable as UTF-8?

aturon replied:
It's akin to http://static.rust-lang.org/doc/master/std/str/fn.from_utf8.html -- basically, there may be cases where you do indeed expect the string to be UTF-8 but you don't want to *panic* if it's not, you want to handle the error more gracefully. By providing an `Option` here we combine the checking and the coercion, a very common pattern in `std`.

alexcrichton commented on src/libstd/ffi/os_str.rs:
```diff
@@ -0,0 +1,214 @@
+    fn eq(&self, other: &OsStr) -> bool {
+        self.bytes().eq(other.bytes())
+    }
+}
```
I suspect that a fairly frequent operation would be to compare an `OsStr` to a `&str` instance, perhaps those impls could be included now as well?

alexcrichton commented on src/libstd/ffi/mod.rs:
```diff
@@ -17,4 +17,8 @@ pub use self::c_str::CString;
 pub use self::c_str::c_str_to_bytes_with_nul;
 
+pub use self::os_str::OsString;
+pub use self::os_str::OsStr;
```
I think this should also export `AsOsStr`

alexcrichton replied:
(not sure if we also want to add `IntoOsString` yet)

also raises questions as to why other conversion traits for `CString` don't exist yet, but they're backwards compatible to add!

alexcrichton commented on src/libstd/ffi/os_str.rs:
```diff
@@ -0,0 +1,214 @@
+        self.bytes().hash(state)
+    }
+}
+
```
Some other traits which may want to be implemented: `Clone`, `Display`, and `Debug`

alexcrichton replied:
er, `Display` would be on a `.display()` adapter most likely

alexcrichton replied:
A few `Add` implementations may also go a long way for showing off how to concatenate `OsStr` instances, or just in general working with OS strings. This may be getting into the weeds though of too many things to implement right off the bat

alexcrichton commented on src/libstd/ffi/os_str.rs:
```diff
@@ -0,0 +1,214 @@
+//!
+//! The need for this type arises from the fact that:
+//!
+//! * On Unix systems, strings are often arbitrary sequences of non-zero
```
This is something that was up for debate on the RFC I believe, but this currently doesn't seem to do any zero-checking anywhere in it. Was that left out by accident, or was that intentional/

alexcrichton commented on src/libstd/ffi/os_str.rs:
```diff
@@ -0,0 +1,214 @@
+pub trait AsOsStr {
+    /// Convert to an `&OsStr` slice.
+    fn as_os_str(&self) -> &OsStr;
+}
```
Do we expect this to be a common trait? I seem to mostly remember `IntoOsStr`, but I can see how this would help `AsPath` at least!

aturon commented:
@alexcrichton Pushed an update addressing most of the issues. I didn't add `Add` just yet; as we discussed today we should probably grow that functionality slowly over time.

Also, as I think I mentioned on IRC, the `IntoOsStr` approach doesn't really work out for io APIs; I expect `AsOsStr` to be the main pattern.

alexcrichton commented on src/libstd/ffi/os_str.rs:
```diff
@@ -0,0 +1,253 @@
+use sys_common::{AsInner, IntoInner, FromInner};
+
+/// Owned, mutable OS strings.
+#[derive(Show, Clone)]
```
Could `Show` be manually implemented instead of derived? (it should delegate to the slice)

alexcrichton commented on src/libstd/sys/unix/os_str.rs:
```diff
@@ -0,0 +1,80 @@
+use string::{String, CowString};
+use mem;
+
+#[derive(Show, Clone)]
```
This `Debug` (`Show`) ad the below `Debug` print out tons of bytes, right? (not necessarily as a `CowString`)

alexcrichton commented on src/libstd/sys/windows/os_str.rs:
```diff
@@ -0,0 +1,76 @@
+use option::Option;
+use mem;
+
+#[derive(Show, Clone)]
```
I think this `Show` should also be manually implemented (to avoid the struct printing around it)

alexcrichton commented:
Just a few nits here and there, but otherwise r=me, can't wait for this to start getting implemented!

aturon commented:
@bors r=alexcrichton e659d55

bors commented:
:hourglass: Testing commit e659d55 with merge 77ecfa3...

bors commented:
:broken_heart: Test failed - [auto-win-64-opt](http://buildbot.rust-lang.org/builders/auto-win-64-opt/builds/1776)

aturon commented:
@bors r=alexcrichton e7850b5

aturon commented:
@bors: r=alexcrichton e7850b5

bors commented:
:hourglass: Testing commit e7850b5 with merge 9d4358a...

bors commented:
:broken_heart: Test failed - [auto-win-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-64-nopt-t/builds/1770)

aturon commented:
@bors: r=alexcrichton dc6cd4c

bors commented:
:hourglass: Testing commit dc6cd4c with merge f14b6ab...

bors commented:
:broken_heart: Test failed - [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/3286)

aturon commented:
@bors: r=alexcrichton 7894490

aturon commented:
@bors: r=alexcrichton c5369eb

bors commented:
:hourglass: Testing commit c5369eb with merge bb7cc4e...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/3284), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/3285), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/3280), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/3288), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/3271), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/3300), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/3296), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/3294), [auto-win-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-32-nopt-t/builds/2952), [auto-win-32-opt](http://buildbot.rust-lang.org/builders/auto-win-32-opt/builds/2954), [auto-win-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-64-nopt-t/builds/1782), [auto-win-64-opt](http://buildbot.rust-lang.org/builders/auto-win-64-opt/builds/1790)
